multicore_runtime/dispatcher.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2015 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""A WSGI app that, once configured, dispatches requests to user apps."""
+
+import logging
+import re
+
+from werkzeug.wrappers import Request
```

Comment:
Ah ok I see. Yes, it turns out imports of the type "from *module* import *Class*, *function*" are common in Python outside of Google and allowed by PEP8, etc.

The Google Python Style Guide suggests a restriction to imports so that only modules get imported. I think the benefit of getting to see the module name everywhere a class or function from that module is used is quite beneficial, as the reader is not hunting that name around until they find it's coming from a module (and modules are usually coarse enough and recognizable, so easy to spot in the code.) The Google Style still conforms to the external guidelines (e.g. PEP8), it's not in violation of anything, it's just an additional restriction.

Let me start a thread about it including other reviewers. In particular I'm not sure if the most appropriate form would be `import werkzeug` (top level), `import werkzeug.wrappers` (using two names) or `from werkzeug import wrappers`. The latter is what the Style Guide's examples indicate, just not sure that the name *wrappers* will give enough context about where this is coming from...